server/src/config.rs

```diff
@@ -0,0 +1,140 @@
+
+use crate::Backend;
+use tower_lsp::lsp_types::{ConfigurationItem, DidChangeConfigurationParams, Url};
+
+#[derive(Default, Debug)]
+pub struct SliceConfig {
+    pub reference_urls: Option<Vec<Url>>,
+}
+
+impl SliceConfig {
+    // Attempt to create a SliceConfig from the backend with the provided uri root.
+    pub async fn try_from_backend(
+        backend: &Backend,
+        root_uri: &Url,
+    ) -> tower_lsp::jsonrpc::Result<SliceConfig> {
+        let reference_directories = Self::fetch_reference_directories(backend).await?;
+        Self::create_config_from_directories(reference_directories, root_uri)
+    }
+
+    // Creates a SliceConfig from the provided configuration parameters and uri root.
+    pub fn from_configuration_parameters(
+        params: DidChangeConfigurationParams,
+        root_uri: &Url,
+    ) -> Self {
+        let reference_directories = Self::parse_reference_directories(&params);
+        Self::create_config_from_directories(reference_directories, root_uri).unwrap_or_default()
+    }
+
+    // Fetch reference directories from the backend.
+    async fn fetch_reference_directories(
+        backend: &Backend,
+    ) -> tower_lsp::jsonrpc::Result<Option<Vec<String>>> {
+        let params = vec![ConfigurationItem {
+            scope_uri: None,
+            section: Some("slice.referenceDirectory".to_string()),
+        }];
+
+        Ok(backend
+            .client
+            .configuration(params)
+            .await?
+            .first()
+            .and_then(|v| v.as_array())
+            .map(|arr| {
+                arr.iter()
+                    .filter_map(|v| v.as_str())
+                    .map(String::from)
+                    .collect::<Vec<_>>()
+            }))
+    }
+
+    // Parse reference directories from configuration parameters.
+    fn parse_reference_directories(params: &DidChangeConfigurationParams) -> Option<Vec<String>> {
+        params
+            .settings
+            .get("slice")
+            .and_then(|v| v.get("referenceDirectory"))
+            .and_then(|v| v.as_array())
+            .map(|arr| {
+                arr.iter()
+                    .filter_map(|v| v.as_str())
+                    .map(String::from)
+                    .collect::<Vec<String>>()
+            })
+    }
+
+    // Create a SliceConfig from the provided reference directories and uri_root.
+    fn create_config_from_directories(
+        reference_directories: Option<Vec<String>>,
+        root_uri: &Url,
+    ) -> Result<SliceConfig, tower_lsp::jsonrpc::Error> {
+        match reference_directories {
+            Some(dirs) => {
+                let reference_urls = Self::try_get_reference_urls(dirs, root_uri)?;
+                Ok(SliceConfig {
+                    reference_urls: Some(reference_urls),
+                })
+            }
+            None => Ok(SliceConfig::default()),
+        }
+    }
+
+    // Convert reference directory strings into URLs.
+    fn try_get_reference_urls(
```

Comment:
I would maybe move some of these functions out side of the impl of SliceConfig and just into the module. I think it would make this module a little easier to follow.